proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+ *
+ * > This section is non-normative.
+ *
+ * We recently added support to Dart Sass that allowed users to opt in to
+ * treating deprecation warnings as errors (on a per-deprecation basis), as
+ * well as opting in early to certain future deprecations. This is currently
+ * supported on the command line and via the Dart API, but we'd like to extend
+ * this support to the JS API as well.
+ *
+ * We would also like to add support for silencing a particular deprecation's
+ * warnings, primarily to enable a gentler process for deprecating `@import`.
+ *
+ * ## Summary
+ *
+ * > This section is non-normative.
+ *
+ * This proposal adds two new interfaces to the JS API (`Deprecation` and
+ * `Version`), three new optional properties on `Options` (`fatalDeprecations`,
+ * `silenceDeprecations`, and `futureDeprecations`), a new parameter on
+ * `Logger.warn` (`options.deprecationType`) two type aliases (`DeprecationOrId`
+ * and `DeprecationStatus`) and a new object `deprecations` that contains the
+ * various `Deprecation` objects.
+ *
+ * All language-wide deprecations are specified in `deprecations`, and any new
+ * deprecations added in the future should update the specification accordingly.
+ * However, the `deprecations` object also allows individual implementations to
+ * add their own, implementation-specific deprecations.
+ *
+ * Every `Deprecation` has a unique `id`, one of four `status` values, and
+ * (optionally) a human-readable `description`. Depending on the status, each
+ * deprecation may also have a `deprecatedIn` version and an `obsoleteIn`
+ * version that specify the compiler versions the deprecation became active
+ * and became obsolete in, respectively.
+ *
+ * ### Design Decisions
+ *
+ * #### Exposing the Full `Deprecation` Interface
+ *
+ * One alternative to specifying a full `Deprecation` interface is to just have
+ * the relevant APIs take in string IDs. We considered this, but concluded that
+ * each deprecation has additional metadata that users of the API may wish to
+ * access (for example, a bundler may wish to surface the `description` and
+ * `deprecatedIn` version to its users).
+ *
+ * #### Formally Specifying the Deprecations
+ *
+ * We chose to make the list of deprecations part of the specification itself,
+ * as this ensures that the language-wide deprecations are consistent across
+ * implementations. However, if an implementation wishes to add a deprecation
+ * that applies only to itself, it may still do so.
+ *
+ * Additionally, while a deprecation's status is part of the specification, we
+ * chose to leave the `deprecatedIn` and `obsoleteIn` versions of each
+ * deprecation out of the specification. As the two current implementers of this
+ * API are both based on Dart Sass, these versions are _currently_ consistent
+ * across implementations in practice, potential future implementers should not
+ * need to be tied to Dart Sass's versioning.
+ *
+ * #### Errors and Warnings for Invalid Deprecations
```

Comment:
We should consider the possibility of users writing configurations that are meant to apply to multiple implementations. Which deprecations have which status is something that's particularly likely to have skew across implementations, so I'm inclined to avoid throwing errors in situations where the user looks like they're passing in more deprecations than necessary to one of these options.